src/models/FileSystem.js

```diff
@@ -4,6 +4,57 @@ import { compareStrings } from '../utils/compareStrings.js'
 import { dirname } from '../utils/dirname.js'
 import { rmRecursive } from '../utils/rmRecursive.js'
 
+function isPromiseFs(fs) {
+  const test = targetFs => {
+    try {
+      // If readFile returns a promise then we can probably assume the other
+      // commands do as well
+      return targetFs.readFile().catch(e => e)
+    } catch (e) {
+      return e
+    }
+  }
+  return test(fs).constructor.name === 'Promise'
+}
+
+// List of commands all filesystems are expected to provide. `rm` is not
+// included since it may not exist and must be handled as a special case
+const commands = [
+  'readFile',
+  'writeFile',
+  'mkdir',
+  'rmdir',
+  'unlink',
+  'stat',
+  'lstat',
+  'readdir',
+  'readlink',
+  'symlink',
+]
+
+function bindFs(target, fs) {
+  if (isPromiseFs(fs)) {
+    for (const command of commands) {
+      target[`_${command}`] = fs[command].bind(fs)
+    }
+  } else {
+    for (const command of commands) {
+      target[`_${command}`] = pify(fs[command].bind(fs))
+    }
+  }
```

Comment:
You don't check if `fs.promises` is provided and always use `pify` that is a fallback for callback-based fs without promises interface in old code.

Reply:
But you can use the `fs` system that has `fs.promises`, but the user passes `fs`, so you use pify to create promises for no reason if there is already `fs.promises`. There are 3 cases here not two.

Reply:
I believe each case is handled correctly no? The execution paths should be:

1. `fs` is passed and `fs.promises` exists
    - `if (promises && promises.enumerable) ` block is triggered
    - `bindfs(fs.promises)` is called
    - `fs.promises` is a promise fs, so `isPromiseFs` evaluates to `true`
    - **commands are not wrapped using `pify`**
2. `fs` is passed and `fs.promises` does not exist, but `fs` is a promise based fs
    - `if (promises && promises.enumerable) ` evaluates to `false`
    - `else` block is triggered
    - `bindfs(fs)` is called
    - `if (isPromiseFs(fs)) ` evaluates to `true`
    - **commands are not wrapped using `pify`**
3. `fs` is passed and `fs.promises` does not exist, and `fs` is not a promise based fs
    -  `if (promises && promises.enumerable) ` evaluates to `false`
    - `else` block is triggered
    - `bindfs(fs)` is called
    - `if(isPromiseFs(fs))` evaluates to `false`
    - **commands are wrapped using `pify`**

If this is incorrect, could you explain the execution path where this does not work, or explain the error in the execution paths I've laid out?

Reply:
> Where is fs.promises.rmdir.bind code?

If `fs.promises` exists, `bindFs(this, fs.promises)` is called, so the `fs` argument to `bindFs` evaluates to `fs.promises`.

Then `if (isPromiseFs(fs))` evaluates to `true`  because `fs.promises`, has been assigned to `fs`, and `fs.promises` is a promise fs.

Since `if (isPromiseFs(fs))` evaluates to `true` then we call `target[`_${command}`] = fs[command].bind(fs)` in a loop.

 `fs.promises` has been assigned to `fs` in the function call, so this statement evaluates to `target[`_${command}`] = fs.promises[command].bind(fs.promises)`

Reply:
Ok, now I understand. Looks good.